### PR TITLE
GEARPUMP-376 Apache Kudu Akka Streaming Sink

### PR DESCRIPTION
**Pull request by SandishKumarHN**
Apache Kudu Akka Streaming Sink.

 - [x] Make sure the commit message is formatted like:
   `[GEARPUMP-376] Meaningful description of pull request` 
 - [x] Make sure tests pass via `sbt clean test`.
 - [x] Make sure old documentation affected by the pull request has been updated and new documentation added for new functionality.

### CODE REVIEW

File: examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/KuduConn.scala [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    val userConfig = new UserConfig(map)
+
+    LOG.info("Kudu userConfig:" + userConfig.getString("kudu.masters"))
+
+
+    val split = new Split
+    val sourceProcessor = DataSourceProcessor(split, splitNum, "Split")
+    val sink = KuduSink(userConfig, "impala::default.kudu_1")
+    LOG.info("KuduSink time")
```
Comment (manuzhang) [inline examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/KuduConn.scala]: Is this logging only for debugging ? Better remove those not providing much "info"

File: examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+class Split extends DataSource {
+
+  private var x = 0
+
+  override def open(context: TaskContext, startTime: Instant): Unit = {}
+
+  override def read(): Message = {
+
+    val tuple = ("column1" -> s"value$x", "column2" -> s"value2$x")
+    x+=1
```
Comment (manuzhang) [inline examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala]: Will `x` ever overflow ?

File: examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/KuduConn.scala [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    val sinkProcessor = DataSinkProcessor(sink, sinkNum)
+    val partitioner = new HashPartitioner
+    val computation = sourceProcessor ~ partitioner ~> sinkProcessor
```
Comment (manuzhang) [inline examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/KuduConn.scala]: Could you please try writing an example with Gearpump's high level `Stream` DSL ? The low level `Processor` API will likely to be hidden from users in the future.

File: external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+  private lazy val kuduWriter = factory.getKuduWriter(userConfig, tableName)
+
+  def this(userConfig: UserConfig, tableName: String) = {
+    this(userConfig, tableName, new KuduWriterFactory)
+  }
+
+  override def open(context: TaskContext): Unit = {}
+
+  // scalastyle:off println
```
Comment (manuzhang) [inline external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala]: no longer used ?

File: external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                  case INT32 => partialRow.addInt(columnName, colValue.toInt)
+                  case INT64 => partialRow.addLong(columnName, colValue.toLong)
+                  case STRING => partialRow.addString(columnName, colValue)
+                  case BOOL => partialRow.addBoolean(columnName, colValue.toBoolean)
+                  case FLOAT => partialRow.addFloat(columnName, colValue.toFloat)
+                  case DOUBLE => partialRow.addDouble(columnName, colValue.toDouble)
+                  case BINARY => partialRow.addByte(columnName, colValue.toByte)
+                  case _ => throw new UnsupportedOperationException(s"Unknown type ${col.getType}")
+                }
+              case _ =>
```
Comment (manuzhang) [inline external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala]: Any warnings here ?

File: external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            }
+          }
+          session.apply(insert)
```
Comment (manuzhang) [inline external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala]: ditto

File: external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        case _ =>
```
Comment (manuzhang) [inline external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala]: better provide some info here

File: external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+import scala.language.implicitConversions
+import org.apache.gearpump.cluster.UserConfig
+import org.apache.gearpump.external.kudu.KuduSink
+import org.apache.gearpump.streaming.dsl.scalaapi.Stream
+
+/** Create a Kudu DSL Sink */
+class KuduDSLSink[T](stream: Stream[T]) {
+
+  def writeToKudu(userConfig: UserConfig, table: String, parallelism: Int, description: String)
```
Comment (manuzhang) [inline external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala]: Please try writing an example with this DSL sink

File: external/kudu/src/test/scala/org/apache/gearpump/external/kudu/KuduSinkSpec.scala [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    )
+    val userConfig = new UserConfig(map)
+    val tableName = "kudu"
+    val key = "key"
+    val value = "value"
+
+    when(kuduClient.openTable(tableName)).thenReturn(table)
+
+    val kuduWriter = new KuduWriter(kuduClient, tableName)
+    // need write test here, get kudu docker image going!
```
Comment (manuzhang) [inline external/kudu/src/test/scala/org/apache/gearpump/external/kudu/KuduSinkSpec.scala]: If it is not yet implemented, we'd better leave it off

File: project/Dependencies.scala [hunk 10]
```diff
@@ -35,6 +35,7 @@ object Dependencies {
   val upickleVersion = "0.3.4"
   val junitVersion = "4.12"
   val kafkaVersion = "0.8.2.1"
+  val kuduversion = "1.7.0"
```
Comment (manuzhang) [inline project/Dependencies.scala]: Please stick to `camelCase` naming

**Comment by manuzhang** [comment]
@SandishKumarHN Thanks for your contribution. I've left some comments.

File: examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala (hunk 2 above)
Comment (SandishKumarHN) [reply examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala]: I just followed HBase example.

File: examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/KuduConn.scala (hunk 3 above)
Comment (SandishKumarHN) [reply examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/KuduConn.scala]: you want me to use KuduSinkDSL?

File: external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala (hunk 4 above)
Comment (SandishKumarHN) [reply external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala]: Will remove

File: external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala (hunk 5 above)
Comment (SandishKumarHN) [reply external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala]: I will just add "Unknown format input format is not supported"

File: external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala (hunk 7 above)
Comment (SandishKumarHN) [reply external/kudu/src/main/scala/org/apache/gearpump/external/kudu/KuduSink.scala]: I will just add "Unknown format input format is not supported"

File: external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala (hunk 8 above)
Comment (SandishKumarHN) [reply external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala]: KuduSinkDSL?

File: external/kudu/src/test/scala/org/apache/gearpump/external/kudu/KuduSinkSpec.scala (hunk 9 above)
Comment (SandishKumarHN) [reply external/kudu/src/test/scala/org/apache/gearpump/external/kudu/KuduSinkSpec.scala]: Okay

File: project/Dependencies.scala (hunk 10 above)
Comment (SandishKumarHN) [reply project/Dependencies.scala]: Okay

File: external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala (hunk 8 above)
Comment (SandishKumarHN) [reply external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala]: can you give a hint on this? how can I call KuduDSLSink?

Comment (manuzhang) [reply external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala]: Please checkout `KafkaReadWrite` example

File: examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala (hunk 2 above)
Comment (manuzhang) [reply examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala]: Then it's an issue for HBase example too

File: external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala (hunk 8 above)
Comment (SandishKumarHN) [reply external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala]: but kudu is sink, it doesn't work as a source.

Kafka can work as source and sink.

  def createAtMostOnceStream[T](
      app: StreamApp,
      topics: String,
      properties: Properties,
      parallelism: Int = 1,
      config: UserConfig = UserConfig.empty,
      description: String = "KafkaSource"
      ): Stream[T] = {
    app.source[T](new KafkaSource(topics, properties), parallelism, config, description)
  }
How would I convert above snippet into kudu one?

File: examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala (hunk 2 above)
Comment (SandishKumarHN) [reply examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala]: what is overflow??

Comment (manuzhang) [reply examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala]: integer overflow

File: external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala (hunk 8 above)
Comment (manuzhang) [reply external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala]: Like `app.source(new Split).sink(new HBaseSink)`

Comment (SandishKumarHN) [reply external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala]: You meant to write a example like "app.source(new Split).sink(new HBaseSink)" for kudusinkdsl??

File: examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala (hunk 2 above)
Comment (SandishKumarHN) [reply examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala]: you want me to put a stop to the integer
like if(x=100000) stop ??

File: external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala (hunk 8 above)
Comment (manuzhang) [reply external/kudu/src/main/scala/org/apache/gearpump/external/kudu/dsl/KuduDSLSink.scala]: Exactly

File: examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala (hunk 2 above)
Comment (manuzhang) [reply examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala]: No, I think `x` needs to be a `Long`

**Comment by manuzhang** [comment]
BTW, please start your commit message with `[GEARPUMP-376]`

File: examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala (hunk 2 above)
Comment (SandishKumarHN) [reply examples/streaming/kudu/src/main/scala/org/apache/gearpump/streaming/examples/kudu/Split.scala]: Done

**Pull request closed without merging**